File: TV/iptvportal_API_examples/jsonsqltest.py
# -*- coding: utf-8 -*-
import os, socket, sys, traceback, urllib2, urlparse, pycurl
try:
    from cStringIO import StringIO
except ImportError:
    from StringIO import StringIO
from simplejson.encoder import JSONEncoder
from simplejson.decoder import JSONDecoder


srv_cfg = {'host': 'admin.go.iptvportal.ru', 'username': 'admin', 'password': 'psw'}


def send (url, data, headers={}):
    method = 'POST'
    timeout = 300
    cacert = None
    c = pycurl.Curl ()
    c.setopt (pycurl.URL, str (url))
    buf = StringIO ()
    c.setopt (pycurl.WRITEFUNCTION, buf.write)
    #c.setopt(pycurl.READFUNCTION, read)
    #data = StringIO (data)
    #c.setopt(pycurl.HEADERFUNCTION, header)
    if cacert:
        c.setopt (c.CAINFO, (cacert)) 
    c.setopt (pycurl.SSL_VERIFYPEER, cacert and 1 or 0)
    c.setopt (pycurl.SSL_VERIFYHOST, cacert and 2 or 0)
    c.setopt (pycurl.ENCODING, 'gzip')
    c.setopt (pycurl.CONNECTTIMEOUT, 30)
    c.setopt (pycurl.TIMEOUT, timeout)
    if method == 'POST':
        c.setopt (pycurl.POST, 1)
        c.setopt (pycurl.POSTFIELDS, data)            
    if headers:
        hdrs = ['%s: %s' % (str(k), str(v)) for k, v in headers.items ()]
        c.setopt (pycurl.HTTPHEADER, hdrs)
    c.perform ()
    c.close ()
    return buf.getvalue ()

_id = 1
def get_id ():
    global _id
    id = _id
    _id += 1
    return id


def jsonrpc_call (uri, method, params, headers={}):
    if type (method) is list:
        methodlist = method
        jsonrpc = []
        for method, params in methodlist:
            jsonrpc.append ({
                'jsonrpc' : "2.0",
                'id'      : get_id (),
                'method'  : method,
                'params'  : params,
            })
    else:
        jsonrpc = {
            'jsonrpc' : "2.0",
            'id'      : get_id (),
            'method'  : method,
            'params'  : params,
        }
    res = send (uri, JSONEncoder (ensure_ascii=True).encode (jsonrpc), headers=headers)
    #print "res:", res
    if not res:
        print "error: not result"
        return None
    res = JSONDecoder (encoding='utf-8').decode (res)
    if type (res) is list:
        return res
    elif not res.get ('result'):
        print "error:" , res.get ('error')
        return False
    else:
        return res.get ('result')


class JsonSqlRpc (object):

    iptvportal_auth_header = None

    def __init__ (self, host=None, username=None, password=None):
        self.host = host
        self.username = username
        self.password = password
        self.jsonrpc_url = 'https://%s/api/jsonrpc/' % host
        self.jsonsql_url = 'https://%s/api/jsonsql/' % host

    def authorize_user (self, username=None, password=None):
        res = jsonrpc_call (self.jsonrpc_url, "authorize_user", {
            'username': username or self.username,
            'password': password or self.password,
        })
        #print "res:", res
        if res: self.iptvportal_auth_header = {'Iptvportal-Authorization': 'sessionid=' + res.get ('session_id')}
        return res

    def jsonsql_call (self, cmd, params=None):
        return jsonrpc_call (self.jsonsql_url, cmd, params, headers=self.iptvportal_auth_header)


srv = JsonSqlRpc (**srv_cfg)
srv_user = srv.authorize_user ()
#print 'authorize srv user:', srv_user


# выборка списка абонентов
res = srv.jsonsql_call ("select", {
    "data": ["username", "password"],
    "from": "subscriber"
})
print 'select cmd result:', res

# выборка списка тв медиа
res = srv.jsonsql_call ("select", {
    "data": ["name", {
        "concat": ["protocol", "://", "inet_addr",
                            {"coalesce": [{"concat": [":", "port"]}, ""]},
                            {"coalesce": [{"concat": ["/", "path"]}, ""]}
        ], "as": "mrl"}],
    "from": "media",
    "where": {"eq": ["is_tv", True]}
})
print 'select cmd result:', res

# выборка списка терминалов
res = srv.jsonsql_call ("select", {
    "data": [{"t": "inet_addr"}, {"t": "mac_addr"}, {"s": "username"}],
    "from": [{
        "table": "terminal", "as": "t"
    }, {
        "join": "subscriber", "join_type": "left", "as": "s", "on": {"eq": [{"t": "subscriber_id"}, {"s": "id"}]}
    }],
    "order_by": {"s": "username"}
})
print 'select cmd result:', res

# добавление абонента "123456" с паролем "111"
res = srv.jsonsql_call ("insert", {
    "into": "subscriber",
    "columns": ["username", "password"],
    "values": {
        "username": "123456",
        "password": "111",
    },
    "returning": "id"
})
print 'insert cmd result:', res

# добавление терминала с мак-адресом '11-22-33-44-55-66' абоненту "123456"
res = srv.jsonsql_call ("insert", {
    "into": "terminal",
    "columns": ["subscriber_id", "mac_addr", "registered"],
    "select": {
        "data": ["id", '11-22-33-44-55-77', True],
        "from": {
            "table": "subscriber", "as": "s"
        },
        "where": {
            "eq": ["username", "123456"]
        }
    },
    "returning": "id"
})
print 'insert cmd result:', res

# добавление пакетов "movie2", "sports2"
res = srv.jsonsql_call ("insert", {
    "into": "package",
    "columns": ["name", "paid"],
    "values": [["movie2", True], ["sports2", True]],
    "returning": "id"
})
print 'insert cmd result:', res

# добавление пакетов "movie", "sports" абоненту "123456"
res = srv.jsonsql_call ("insert", {
    "into": "subscriber_package",
    "columns": ["subscriber_id", "package_id", "enabled"],
    "select": {
        "data": [{"s": "id"}, {"p": "id"}, True],
        "from": [{
            "table": "subscriber", "as": "s"
        }, {
            "table": "package", "as": "p"
        }],
        "where": {
            "and": [{
                "eq": [{"s": "username"}, "123456"]
            }, {
                "in": [{"p": "name"}, "movie", "sports"]
            }]
        }
    },
    "returning": "package_id"
})
print 'insert cmd result:', res

# отключение абонента с акаунтом "123456"
res = srv.jsonsql_call ("update", {
    "table": "subscriber",
    "set"  : {
        #"disabled": {'not': {'subscriber': "disabled"}}
        "disabled": {'not': "disabled"}
    },
    "where": {"eq": ["username", "123456"]},
    "returning": "id"
})
print 'update cmd result:', res

# удаление абонентских устройств акаунта "123456"
res = srv.jsonsql_call ("delete", {
    "from": "terminal",
    "where": {"in": ["subscriber_id", {
        "select": {
            "data": "id",
            "from": "subscriber",
            "where": {"eq": ["username", "123456"]}
        }
    }]},
    "returning": "id"
})
print 'delete cmd result:', res

# удаление пакетов для акаунта "123456"
res = srv.jsonsql_call ("delete", {
    "from": "subscriber_package",
    "where": {"in": ["subscriber_id", {
        "select": {
            "data": "id",
            "from": "subscriber",
            "where": {"eq": ["username", "123456"]}
        }
    }]},
    "returning": "package_id"
})
print 'delete cmd result:', res

# список запросов
res = srv.jsonsql_call ([(
    "select", {
    "data": ["name", {
        "concat": ["protocol", "://", "inet_addr",
                            {"coalesce": [{"concat": [":", "port"]}, ""]},
                            {"coalesce": [{"concat": ["/", "path"]}, ""]}
        ], "as": "mrl"}],
    "from": "media",
    "where": {
        "and": [{
            "eq": ["is_tv", True]
        }, {
            "eq": ["inet_addr", "235.10.10.1"]
        }]
    }}), (
    "select", {
    "data": ["name", {
        "concat": ["protocol", "://", "inet_addr",
                            {"coalesce": [{"concat": [":", "port"]}, ""]},
                            {"coalesce": [{"concat": ["/", "path"]}, ""]}
        ], "as": "mrl"}],
    "from": "media",
    "where": {
        "and": [{
            "eq": ["is_tv", True]
        }, {
            "eq": ["inet_addr", "235.10.10.2"]
        }]
    }}), (
    "select", {
    "data": ["name", {
        "concat": ["protocol", "://", "inet_addr",
                            {"coalesce": [{"concat": [":", "port"]}, ""]},
                            {"coalesce": [{"concat": ["/", "path"]}, ""]}
        ], "as": "mrl"}],
    "from": "media",
    "where": {
        "and": [{
            "eq": ["is_tv", True]
        }, {
            "eq": ["inet_addr", "235.10.10.3"]
        }]
    }})
], None)
print 'select cmd result:', res


# выборка списка значений
res = srv.jsonsql_call ("select", {
    "values": [[1, 'a'], [2, 'b'], [3, 'c']],
})
print 'select cmd result:', res

# выборка списка значений
res = srv.jsonsql_call ("select", {
    "data": ["username", "password"],
    "from": "subscriber",
    "where": {
        "in": ["username", {"values": [['12345'], ['123456']]}]
    }
})
print 'select cmd result:', res

File: TV/iptvportal_API_examples/jsonsqltest.php
<?php

$_auth_uri = 'https://go.iptvportal.ru/api/jsonrpc/';
$_username = 'admin'; $_password = 'psw';
$_jsonsql_uri = 'https://192.168.0.250/api/jsonsql/';
$_iptvportal_header = null;

function send ($url, $data, $extra_headers=null) {
    $ch = curl_init ();
    curl_setopt ($ch, CURLOPT_CONNECTTIMEOUT,  10);
    curl_setopt ($ch, CURLOPT_RETURNTRANSFER,  true);
    curl_setopt ($ch, CURLOPT_TIMEOUT,         10);
    curl_setopt ($ch, CURLOPT_URL,             $url);
    curl_setopt ($ch, CURLOPT_SSL_VERIFYPEER,  false);
    curl_setopt ($ch, CURLOPT_SSL_VERIFYHOST,  false);
    curl_setopt ($ch, CURLOPT_POST,            true);
    curl_setopt ($ch, CURLOPT_POSTFIELDS,      $data);
    if (isset ($extra_headers)) {
        curl_setopt ($ch, CURLOPT_HTTPHEADER,  $extra_headers);
    }
    //echo "HTTP fetching '$url'...\n";
    $content = curl_exec ($ch);
    $http_code = curl_getinfo ($ch, CURLINFO_HTTP_CODE);
    if ($content === false) {
        $err_msg = "HTTP error: $http_code (" . curl_error($ch) . ')' . '\n';
        echo $err_msg;
        throw new Exception ($err_msg);
    }
    if ($http_code != 200) {
        $err_msg = "HTTP request failed ($http_code)\n";
        echo $err_msg;
        throw new Exception ($err_msg);
    }
    //echo "HTTP OK ($http_code)\n";
    curl_close ($ch);
    return $content;
}

function jsonrpc_call ($url, $method, $params, $extra_headers=null) {
    static $req_id = 1;

    $req = array (
        "jsonrpc" => '2.0',
        "id" => $req_id++,
        "method" => $method,
        "params" => $params
    );
    $req = json_encode ($req);
    $res = send ($url, $req, $extra_headers=$extra_headers);
    #echo $res;
    $res = json_decode ($res, true);
    if (!isset ($res)) {
        echo "error: not result\n";
        return null;
    } else if (!array_key_exists ('result', $res) || !isset ($res ['result'])) {
        print_r ($res ['error']);
        return null;
    } else {
        return $res ['result'];
    }
    return $res;
}

function jsonsql_call ($cmd, $params) {
    global $_jsonsql_uri, $_iptvportal_header;
    //echo 'iptvportal_header: '; print_r ($_iptvportal_header);
    return jsonrpc_call ($_jsonsql_uri, $cmd, $params, $extra_headers=$_iptvportal_header);
}

function authorize_user ($auth_uri, $username, $password) {
    global $_iptvportal_header;
    $res = jsonrpc_call ($auth_uri, $cmd="authorize_user", $params=array (
        'username' => $username,
        'password' => $password
    ));
    if (isset ($res) && array_key_exists ('session_id', $res)) {
        $_iptvportal_header = array ('Iptvportal-Authorization: ' . 'sessionid=' . $res ['session_id']);
    }
    return $res;
}

$user = authorize_user ($auth_uri=$_auth_uri, $username=$_username, $password=$_password);
echo 'authorize user result: '; print_r ($user);

# выборка списка абонентов
$res = jsonsql_call ("select", array (
    "data" => array ("username", "password"),
    "from" => "subscriber"
));
echo 'select cmd result: '; print_r ($res);

# выборка списка тв медиа
$res = jsonsql_call ("select", array (
    "data" => array ("name", array (
        "concat" => array ("protocol", "://", "inet_addr",
                            array ("coalesce" => array (array ("concat" => array (":", "port")), "")),
                            array ("coalesce" => array (array ("concat" => array ("/", "path")), ""))
        ), "as" => "mrl")),
    "from" => "media",
    "where" => array ("eq" => array ("is_tv", true))
));
#echo 'select cmd result: '; print_r ($res);

# выборка списка терминалов
$res = jsonsql_call ("select", array (
    "data" => array (array ("t" => "inet_addr"), array ("t" => "mac_addr"), array ("s" => "username")),
    "from" => array (
        array ("table" => "terminal", "as" => "t"),
        array ("join" => "subscriber", "join_type" => "left", "as" => "s",
               "on" => array ("eq" => array (array ("t" => "subscriber_id"), array ("s" => "id")))
        )
    ),
    "order_by" => array ("s" => "username")
));
echo 'select cmd result: '; print_r ($res);

# добавление абонента "123456" с паролем "111"
$res = jsonsql_call ("insert", array (
    "into" => "subscriber",
    "columns" => array ("username", "password"),
    "values" => array (
        "username" => "123456",
        "password" => "111",
    ),
    "returning" => "id"
));
echo 'insert cmd result: '; print_r ($res);

# добавление терминала с мак-адресом '11-22-33-44-55-66' абоненту "123456"
$res = jsonsql_call ("insert", array (
    "into" => "terminal",
    "columns" => array ("subscriber_id", "mac_addr", "registered"),
    "select" => array (
        "data" => array ("id", '11-22-33-44-55-66', true),
        "from" => array (
            "table" => "subscriber", "as" => "s"
        ),
        "where" => array (
            "eq" => array ("username", "123456")
        )
    ),
    "returning" => "id"
));
echo 'insert cmd result: '; print_r ($res);

# добавление пакетов "movie", "sports" абоненту "123456"
$res = jsonsql_call ("insert", array (
    "into" => "subscriber_package",
    "columns" => array ("subscriber_id", "package_id", "enabled"),
    "select" => array (
        "data" => array (array ("s" => "id"), array ("p" => "id"), true),
        "from" => array (array (
            "table" => "subscriber", "as" => "s"
        ), array (
            "table" => "package", "as" => "p"
        )),
        "where" => array (
            "and" => array (array (
                "eq" => array (array ("s" => "username"), "123456")
            ), array (
                "in" => array (array ("p" => "name"), "movie", "sports")
            ))
        )
    ),
    "returning" => "package_id"
));
echo 'insert cmd result: '; print_r ($res);

# получение пакетов для акаунта "123456"
$res = jsonsql_call ("select", array (
    "data" => array (array ("p" => "id"), array ("p" => "name")),
    "from" => array (
        array ("table" => "package", "as" => "p"),
        array ("join" => "subscriber_package", "join_type" => "inner", "as" => "s2p",
               "on" => array ("eq" => array (array ("s2p" => "package_id"), array ("p" => "id")))
        ),
        array ("join" => "subscriber", "join_type" => "inner", "as" => "s",
               "on" => array ("eq" => array (array ("s2p" => "subscriber_id"), array ("s" => "id")))
        )
    ),
    "where" => array ("eq" => array (array ("s" => "username"), "123456")),
    "order_by" => array ("p" => "name")
));
echo 'select cmd result: '; print_r ($res);

# получение терминалов для акаунта "123456"
$res = jsonsql_call ("select", array (
    "data" => array (array ("t" => "id"), array ("t" => "inet_addr"), array ("t" => "mac_addr")),
    "from" => array (
        array ("table" => "terminal", "as" => "t"),
        array ("join" => "subscriber", "join_type" => "inner", "as" => "s",
               "on" => array ("eq" => array (array ("t" => "subscriber_id"), array ("s" => "id")))
        )
    ),
    "where" => array ("eq" => array (array ("s" => "username"), "123456"))
));
echo 'select cmd result: '; print_r ($res);

# отключение абонента с акаунтом "123456"
$res = jsonsql_call ("update", array (
    "table" => "subscriber",
    "set" => array (
        "disabled" => true
    ),
    "where" => array ("eq" => array ("username", "123456")),
    "returning" => "id"
));
echo 'update cmd result: '; print_r ($res);

# обновляем абонента с акаунтом "123456"
$res = jsonsql_call ("update", array (
    "table" => "subscriber",
    "set" => array (
        "disabled" => true
    ),
    "where" => array ("eq" => array ("username", "123456")),
    "returning" => "id"
));
echo 'update cmd result: '; print_r ($res);
if (!(count ($res) > 0)) {
    # добавление абонента "123456" с паролем "111"
    $res = jsonsql_call ("insert", array (
        "into" => "subscriber",
        "columns" => array ("username", "password"),
        "values" => array (
            "username" => "123456",
            "password" => "111",
        ),
        "returning" => "id"
    ));
    echo 'insert cmd result: '; print_r ($res);
}

# удаление абонентских устройств акаунта "123456"
$res = jsonsql_call ("delete", array (
    "from" => "terminal",
    "where" => array ("in" => array ("subscriber_id", array (
        "select" => array (
            "data" => "id",
            "from" => "subscriber",
            "where" => array ("eq" => array ("username", "123456"))
        )
    ))),
    "returning" => "id"
));
echo 'delete cmd result: '; print_r ($res);

# удаление пакетов "movie", "sports" для акаунта "123456"
$res = jsonsql_call ("delete", array (
    "from" => "subscriber_package",
    "where" => array ("and" => array (
        array ("in" => array ("subscriber_id", array (
            "select" => array (
                "data" => "id",
                "from" => "subscriber",
                "where" => array ("eq" => array ("username", "123456"))
            )
        ))), array ("in" => array ("package_id", array (
            "select" => array (
                "data" => "id",
                "from" => "package",
                "where" => array ("in" => array ("name", "movie", "sports"))
            )
        )))
    )),
    "returning" => "package_id"
));
echo 'delete cmd result: '; print_r ($res);

# удаление пакетов для акаунта "123456"
$res = jsonsql_call ("delete", array (
    "from" => "subscriber_package",
    "where" => array ("in" => array ("subscriber_id", array (
        "select" => array (
            "data" => "id",
            "from" => "subscriber",
            "where" => array ("eq" => array ("username", "123456"))
        )
    ))),
    "returning" => "package_id"
));
echo 'delete cmd result: '; print_r ($res);


?>
